subsys/bluetooth/host/smp.c

```diff
@@ -5663,10 +5667,25 @@ int bt_smp_le_oob_generate_sc_data(struct bt_le_oob_sc_data *le_sc_oob)
 	}
 
 	if (!sc_public_key) {
-		err = k_sem_take(&sc_local_pkey_ready, K_FOREVER);
+		err = k_sem_take(&sc_local_pkey_ready, K_NO_WAIT);
+		if (!err) {
+			/* Public key request has finished, but generation failed. Retrying. */
+			err = bt_pub_key_gen(&sc_local_pub_key_cb);
+			if (err && err != -EALREADY) {
+				return err;
+			}
+		}
+
+		/* Public key has been requested but not ready yet. */
+		err = k_sem_take(&sc_local_pkey_ready, K_SECONDS(30));
```

Comment:
Should this use `BT_ATT_TIMEOUT`?
```suggestion
		err = k_sem_take(&sc_local_pkey_ready, BT_ATT_TIMEOUT);
```

Reply:
Ah, right. Somehow I had remember that SMP was on ATT and not on L2CAP :)